File: Java/src/data/stack/Stack.java
package data.stack;

import data.NodeItem;
import java.util.NoSuchElementException;

// Stacks are last in first out (LIFO) data structures
public class Stack<T> {
  private NodeItem<T> top;

  /**
   * Remove the last item in the stack
   * @return T - Last item that was put in the stack
   */
  public T pop() {
    if (top == null) {
      throw new NoSuchElementException();
    }

    T item = top.data;
    top = top.next;
    return item;
  }

  /**
   * Adds item to end of stack
   * @param T Item
   */
  public void push(T item) {
    NodeItem<T> t = new NodeItem<T>(item);
    t.next = top;
    top = t;
  }

  /**
   * Look at the last item in the stack
   * @return T Item
   */
  public T peek() {
    if (top == null) {
      throw new NoSuchElementException();
    }

    return top.data;
  }

  /**
   * Check if stack is empty
   * @return Boolean - True if stack is empty else false
   */
  public boolean isEmpty() {
    return top == null;
  }
}

File: JavaScript/data_structures/stacks/Stack.js
// Stack Data Structure

// Stacks serve as a collection of elements using the LIFO (last in first out) ordering.

// An array in JavaScript has built-in methods for using it as as stack and it may not
// be necessary to create a Stack object. This is done here for demonstrative purposes.

class Stack {
    constructor() {
        this.stack = [];
    }

    // Push an element onto the top of the stack.
    push(value) {
        this.stack.push(value);
    }

    // Remove and return the element at the top of the stack.
    pop() {
        return this.stack.pop();
    }

    // Return the element at the top of the stack but don't remove
    // it from the stack.
    peek() {
        return this.isEmpty() ? null : this.stack[this.stack.length -1];
    }

    // Check if the stack is empty.
    isEmpty() {
        return this.stack.length === 0;
    }
}

module.exports = Stack;